blog/views.py

```diff
@@ -1,117 +1,127 @@
-    """Publish post."""
-    post = get_object_or_404(Post, pk=pk)
-    post.publish()
-    return redirect('post_detail', pk=pk)
-
-
-@login_required
-def post_remove(request, pk):
-    """Delete post."""
-    post = get_object_or_404(Post, pk=pk)
-    post.delete()
-    return redirect('post_list')
-
-
-def add_comment(request, pk):
-    """Add comment to post."""
-    post = get_object_or_404(Post, pk=pk)
-    if request.method == 'POST':
-        form = CommentForm(request.POST)
-        if form.is_valid():
-            comment = form.save(commit=False)
-            comment.post = post
-            comment.save()
-            return redirect('post_detail', pk=post.pk)
-    else:
-        form = CommentForm()
-    return render(request, 'blog/add_comment.html', {'form': form})
-
-
-@login_required
-def comment_approve(request, pk):
-    """Approve (publish) comment."""
-    comment = get_object_or_404(Comment, pk=pk)
-    comment.approve()
-    return redirect('post_detail', pk=comment.post.pk)
-
-
-@login_required
-def comment_remove(request, pk):
-    """Delete comment."""
-    comment = get_object_or_404(Comment, pk=pk)
-    post_pk = comment.post.pk
-    comment.delete()
-    return redirect('post_detail', pk=post_pk)
+class Protected(View):
+    """Deny access to view for unauthorized users."""
+
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Call base method with login_required extension."""
+        return super().dispatch(*args, **kwargs)
+
+
+class PostListView(ListView):
+    """View for list of all published posts."""
+
+    context_object_name = 'posts'
+    template_name = 'blog/post_list.html'
+
+    def get_queryset(self):
+        """Return already published posts."""
+        return Post.objects.filter(published_date__lte=timezone.now()).order_by('published_date')
+
+
+class PostDetailView(DetailView):
+    """View with detail info about post."""
+
+    model = Post
+    template_name = 'blog/post_detail.html'
+
+
+class NewPost(CreateView, Protected):
+    """View for creating new post."""
+
+    form_class = PostForm
+    template_name = 'blog/post_edit.html'
+
+    def form_valid(self, form):
+        """Pass request.user to model."""
+        form.save(self.request.user)
+        return super().form_valid(form)
+
+
+class EditPost(UpdateView, Protected):
+    """View for editing post."""
+
+    model = Post
+    fields = ['title', 'text']
+    template_name = 'blog/post_edit.html'
+
+
+class PostDraftList(ListView, Protected):
+    """View for list with unpublished posts."""
+
+    context_object_name = 'posts'
+    template_name = 'blog/post_draft_list.html'
+    queryset = Post.objects.filter(published_date__isnull=True).order_by('created_date')
+
+
+class PublishPost(UpdateView, Protected):
+    """View for publishing post."""
+
+    model = Post
+    fields = ['title', 'text']
+    template_name = 'blog/post_detail.html'
+
+    def get(self, request, *args, **kwargs):
```

Comment:
That's weird: why not just `POST` the change and let update view to handle all internal logic?

Reply:
Should I remove publish() method from model then? And just pass published_date change?

Reply:
no, that method is handy